File: firebase/functions/lib/events/live_meetings/breakouts/get_breakout_room_assignment.dart
import 'dart:async';

import 'package:firebase_admin_interop/firebase_admin_interop.dart';
import 'package:firebase_functions_interop/firebase_functions_interop.dart';
import '../../../on_call_function.dart';
import '../../../utils/infra/firestore_utils.dart';
import 'package:data_models/cloud_functions/requests.dart';
import 'package:data_models/events/event.dart';
import 'package:data_models/events/live_meetings/live_meeting.dart';
import 'package:data_models/utils/utils.dart';

class GetBreakoutRoomAssignment
    extends OnCallMethod<GetBreakoutRoomAssignmentRequest> {
  GetBreakoutRoomAssignment()
      : super(
          'GetBreakoutRoomAssignment',
          (jsonMap) => GetBreakoutRoomAssignmentRequest.fromJson(jsonMap),
        );

  @override
  Future<Map<String, dynamic>> action(
    GetBreakoutRoomAssignmentRequest request,
    CallableContext context,
  ) async {
    // Verify user is a participant
    final participant = await firestoreUtils.getFirestoreObject(
      path: '${request.eventPath}/event-participants/${context.authUid}',
      constructor: (map) => Participant.fromJson(map),
    );
    if (participant.status != ParticipantStatus.active) {
      throw HttpsError(HttpsError.failedPrecondition, 'unauthorized', null);
    }

    final liveMeetingPath =
        '${request.eventPath}/live-meetings/${request.eventId}';
    final liveMeeting = await firestoreUtils.getFirestoreObject(
      path: liveMeetingPath,
      constructor: (map) => LiveMeeting.fromJson(map),
    );

    final breakoutSession = liveMeeting.currentBreakoutSession;

    if (breakoutSession == null ||
        breakoutSession.breakoutRoomStatus != BreakoutRoomStatus.active) {
      throw Exception('Breakout rooms not active');
    }

    final breakoutRoomsCollection = firestore.collection(
      '$liveMeetingPath/breakout-room-sessions/${breakoutSession.breakoutRoomSessionId}/breakout-rooms',
    );

    final currentBreakoutRoom = await breakoutRoomsCollection
        .where(
          BreakoutRoom.kFieldParticipantIds,
          arrayContains: context.authUid,
        )
        .get();

    String? assignment;
    if (currentBreakoutRoom.documents.isNotEmpty) {
      assignment = BreakoutRoom.fromJson(
        firestoreUtils.fromFirestoreJson(
          currentBreakoutRoom.documents.first.data.toMap(),
        ),
      ).roomId;
    } else {
      final allBreakoutRoomDocs = await breakoutRoomsCollection.get();
      final allBreakoutRooms = allBreakoutRoomDocs.documents
          .map(
            (doc) => BreakoutRoom.fromJson(
              firestoreUtils.fromFirestoreJson(doc.data.toMap()),
            ),
          )
          .toList();

      assignment = await firestore.runTransaction((transaction) async {
        // Assign to waiting room if there is one.
        var assignedRoomIndex = allBreakoutRooms
            .indexWhere((br) => br.roomId == breakoutsWaitingRoomId);
        if (assignedRoomIndex < 0 && allBreakoutRooms.isNotEmpty) {
          // If no waiting room, then assign to a room with least participants
          final minBreakoutRoomCount = allBreakoutRooms
              .map((b) => b.participantIds.length)
              .reduce((a, b) => a < b ? a : b);
          assignedRoomIndex = allBreakoutRooms.indexWhere(
            (b) => b.participantIds.length == minBreakoutRoomCount,
          );
        }

        if (assignedRoomIndex < 0) {
          print('error, no breakout rooms found.');
          return null;
        } else {
          final roomDoc = breakoutRoomsCollection
              .document(allBreakoutRooms[assignedRoomIndex].roomId);
          final assignedRoomLookup = await firestoreUtils.getFirestoreObject(
            transaction: transaction,
            path: roomDoc.path,
            constructor: (map) => BreakoutRoom.fromJson(map),
          );

          transaction.update(
            roomDoc,
            UpdateData.fromMap(
              jsonSubset(
                [BreakoutRoom.kFieldParticipantIds],
                firestoreUtils.toFirestoreJson(
                  assignedRoomLookup.copyWith(
                    participantIds: [
                      ...assignedRoomLookup.participantIds,
                      context.authUid!,
                    ],
                  ).toJson(),
                ),
              ),
            ),
          );

          return assignedRoomLookup.roomId;
        }
      });
    }

    return GetBreakoutRoomAssignmentResponse(roomId: assignment).toJson();
  }
}

File: firebase/functions/src/events/live_meetings/breakouts/get_breakout_room_assignment.ts
import * as functions from 'firebase-functions';
import * as admin from 'firebase-admin';
import { OnCallMethod } from '../../../on_call_function';
import { firestore, firestoreUtils } from '../../../utils/infra/firestore_utils';
import { Participant, ParticipantStatus, LiveMeeting, BreakoutRoom, BreakoutRoomSession, BreakoutRoomStatus, breakoutsWaitingRoomId } from '../../../types';

interface GetBreakoutRoomAssignmentRequest {
  eventPath: string;
  eventId: string;
}

interface GetBreakoutRoomAssignmentResponse {
  breakoutRoomId?: string;
}

export class GetBreakoutRoomAssignment extends OnCallMethod<GetBreakoutRoomAssignmentRequest> {
  constructor() {
    super('GetBreakoutRoomAssignment', (jsonMap) => jsonMap as GetBreakoutRoomAssignmentRequest);
  }

  async action(
    request: GetBreakoutRoomAssignmentRequest,
    context: functions.https.CallableContext
  ): Promise<GetBreakoutRoomAssignmentResponse> {
    const participant: Participant = await firestoreUtils.getFirestoreObject({
      path: `${request.eventPath}/event-participants/${context.auth?.uid}`,
      constructor: (map) => map as unknown as Participant,
    });
    if ((participant as any).status !== ParticipantStatus.active) {
      throw new functions.https.HttpsError('failed-precondition', 'unauthorized');
    }

    const liveMeetingPath = `${request.eventPath}/live-meetings/${request.eventId}`;
    const liveMeeting: LiveMeeting = await firestoreUtils.getFirestoreObject({
      path: liveMeetingPath,
      constructor: (map) => map as unknown as LiveMeeting,
    });

    const breakoutSession = (liveMeeting as any).currentBreakoutSession as BreakoutRoomSession | undefined;

    if (!breakoutSession || breakoutSession.breakoutRoomStatus !== BreakoutRoomStatus.active) {
      throw new Error('Breakout rooms not active');
    }

    const breakoutRoomsPath = `${liveMeetingPath}/breakout-room-sessions/${breakoutSession.breakoutRoomSessionId}/breakout-rooms`;
    const breakoutRoomsCollection = firestore.collection(breakoutRoomsPath);

    const currentBreakoutRoomQuery = await breakoutRoomsCollection
      .where('participantIds', 'array-contains', context.auth?.uid)
      .get();

    if (!currentBreakoutRoomQuery.empty) {
      const room = firestoreUtils.fromFirestoreJson(currentBreakoutRoomQuery.docs[0].data()) as unknown as BreakoutRoom;
      return { breakoutRoomId: (room as any).roomId };
    }

    const allBreakoutRoomDocs = await breakoutRoomsCollection.get();
    const allBreakoutRooms = allBreakoutRoomDocs.docs.map((doc) =>
      firestoreUtils.fromFirestoreJson(doc.data()) as unknown as BreakoutRoom
    );

    const assignment = await firestore.runTransaction(async (transaction) => {
      let assignedRoomIndex = allBreakoutRooms.findIndex((br) => (br as any).roomId === breakoutsWaitingRoomId);
      if (assignedRoomIndex < 0 && allBreakoutRooms.length > 0) {
        const minCount = Math.min(...allBreakoutRooms.map((b) => ((b as any).participantIds ?? []).length));
        assignedRoomIndex = allBreakoutRooms.findIndex((b) => ((b as any).participantIds ?? []).length === minCount);
      }

      if (assignedRoomIndex < 0) {
        console.log('error, no breakout rooms found.');
        return null;
      }

      const roomDoc = breakoutRoomsCollection.doc((allBreakoutRooms[assignedRoomIndex] as any).roomId);
      const assignedRoomSnap = await transaction.get(roomDoc);
      const assignedRoom = firestoreUtils.fromFirestoreJson(assignedRoomSnap.data() ?? {}) as unknown as BreakoutRoom;
      const participantIds = [...((assignedRoom as any).participantIds ?? []), context.auth!.uid];

      transaction.update(roomDoc, { participantIds });
      return (assignedRoom as any).roomId as string;
    });

    return { breakoutRoomId: assignment ?? undefined };
  }
}
